Add in key remapping for improved sort merge join performance

revans2 opened the pull request:
## Description
This PR adds in APIs for join key remapping. Most of the code here was written by AI, but based on the existing distinct join code.  This is very close to a left outer distinct join, but it adds in a few changes.  

1. There are special sentinel values when something is not found. This lets us deal with null equality properly when remapping keys
2. There is built in metrics calculation.  This computes the distinct count of the keys and the the number of key collisions that there would be. This is because those two values are what I found correlate best to selecting the proper join type. If all the keys are distinct then we want to do a distinct join. If there are a low number of key collisions, then hash join always wins. If there is a high number, then sort merge join wins.

In fact because this is so close to distinct join, I was thinking that we might want to just combine it with distinct join, so if it does end up that we want to do a distinct join, then the join hash table is already created.  I have kept is separate for now just to keep things simpler.

If this looks good I am happy to make whatever changes are needed to get this production ready. I do not have benchmark data yet. I keep running into some issues with the benchmarks (even running existing ones) that I am trying to work through.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

abellina commented on java/src/test/java/ai/rapids/cudf/KeyRemappingTest.java:
```diff
@@ -0,0 +1,154 @@
+  @Test
+  void testStringKeys() {
+    try (Table buildTable = new Table.TestBuilder()
+            .column("apple", "banana", "cherry", "banana")
```
I am also curious if we should add a key remapping determinism test, where we run it several times and sort and verify we get the same remappings.

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,641 @@
+
+    // Determine if we need to filter nulls
+    bool const skip_nulls =
+      (_compare_nulls == cudf::null_equality::UNEQUAL) && cudf::nullable(build);
```
Use `table_view::has_nested_nulls` for checking nested columns.

revans2 replied:
But that is not what Spark does, and it will not work properly for lists. I don't think that null equality is applied equally in all cases.

ttnghia commented on cpp/include/cudf/join/key_remapping.hpp:
```diff
@@ -0,0 +1,134 @@
+   *
+   * For each row in the input, returns the integer ID assigned to that key.
+   * - Keys that match a build table key: return a non-negative integer
+   * - Keys with nulls (when compare_nulls is UNEQUAL): return KEY_REMAP_BUILD_NULL
```
What is the output for null keys when null compared `EQUAL`?

revans2 replied:
It would be the same for any other key. In this case it would be the index of one of the rows with a null in it, but the contract just says that it would be a unique positive integer.

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,651 @@
+
+    // Atomically increment the count for this key's first occurrence
+    // This is a distributed atomic (different keys update different locations)
+    atomicAdd(&counts_ptr[stored_idx], 1);
```
If we exclude nulls from the hash table insertion (for `insert_and_count_with_nulls_fn`) then we have to do some extra work:
```
** Insert kernel:
if (!cudf::bit_is_set(bitmask_ptr, idx)) { 
  insert_markers[idx] = key_insert_marker::NULL;
  inserted_keys[idx] = -1;
  return;
}

** Count duplicate
auto const count_one_it = thrust::make_transform_iterator(
  thrust::make_counting_iterator(0),
  [insert_markers = insert_markers.begin()]__device__(auto const idx) -> size_type { 
    // If the key is null then always return `0` for each occurrence.
    return insert_markers[idx] == key_insert_marker::NULL : 0 : 1
  });

thrust::reduce_by_key( 
    sorted_key_it, sorted_key_it + num_keys, 
    count_one_it,
    thrust::make_discard_iterator(), // discard count for keys
    key_counts.begin());
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,651 @@
+ * @brief Functor for inserting keys with counting (with null filtering).
+ */
+template <typename SetRef, typename KeyIter>
+struct insert_and_count_with_nulls_fn {
```
Indeed the functor `insert_and_count_fn` below is unnecessary. We can just use this same implementation for both cases (with nulls and non-nulls). For the non-nulls cases, just construct the functor with a `nullptr` value for the bitmask. Then, modify the statement:
```
if (!cudf::bit_is_set(bitmask_ptr, idx)) { return; }
```
into this:
```
if (bitmask_ptr && !cudf::bit_is_set(bitmask_ptr, idx)) { return; }
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,651 @@
+    }
+  }
+
+  bool _has_nested_columns;
```
```suggestion
  bool _build_has_nested_columns;
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,651 @@
+                     cudf::null_equality compare_nulls,
+                     rmm::cuda_stream_view stream)
+    : _compare_nulls{compare_nulls},
+      _table{create_key_remap_table(build, compare_nulls, 0.5, stream)}
```
Again, remove this hard coded load factor and use a global `constexpr` constant only when we call cuco API.

revans2 commented:
@ttnghia I think I have addressed all of your review comments except for the sort based key remapping. If you look back at the history I did add in the sort based key remapping along with some benchmarks to help me understand if it is better than hash and in what ways. As a note I updated the hash based remapping to act more like the current group by aggregations on a count. I also made the metrics calculation optional so that if someone does not want the metrics, or if they were already calculated and cached, they do not need to be calculated again.

Also as a note I will be putting up a PR shortly to help fix some issues with sort_merge_join.cu. If the output number of rows would be larger that 2^32 we could get overflow problems and end up with out of bounds memory accesses. So the benchmarks are run with that fix, just so that I could better see across the entire range of values.

<img width="1665" height="1186" alt="heuristics_heatmap" src="https://github.com/user-attachments/assets/f4424c32-e97f-4f5f-8756-617528a0fd75" />

This shows that the new hash based key remapping is always a win over the naive group by with a count, followed by a max reduction. This is likely because we skip gathering the keys.

<img width="1916" height="1179" alt="algorithm_comparison" src="https://github.com/user-attachments/assets/da7bff8a-34d7-4445-8d28-3bc0e57553a9" />

This shows the comparison between the sort based vs the hash based on several different levels of cardinality and data types. Effectively the only time that the sort remapping wins is when we have a very skewed number of duplicate keys. This is also the case when a sort merge join wins over a hash join. So then I wanted to see what percentage of the time was spent doing the join vs calculating the heuristic.

<img width="2384" height="1776" alt="time_breakdown_grid" src="https://github.com/user-attachments/assets/50b23b71-e53b-4cc0-b6a2-44b0ff473809" />

From this you can see that the cases when the the sort reduce metrics calculation wins are the ones when the join is the slowest and the metrics calculation is less important in the overall runtime of the query. As such I decided that it would likely be better overall to use the hash based/atomic metrics calculations. But I am open to a different interpretation as memory requirements for the joins increased with fewer distinct keys, which then resulted in me only being able to run fewer and fewer join tests, so there could be some skew in the aggregate results I am presenting.

ttnghia commented on cpp/benchmarks/join/join_on_int32.cu:
```diff
@@ -0,0 +1,164 @@
+ */
+
+// Cardinality distributions (same as key_remap_build)
+enum class Cardinality { ALL_UNIQUE, HIGH_UNIQUE, MED_UNIQUE, LOW_UNIQUE, SINGLE_KEY };
```
Style convention in cudf is using `snake_case` for enum class name.
```suggestion
enum class key_cardinality { ALL_UNIQUE, HIGH_UNIQUE, MED_UNIQUE, LOW_UNIQUE, SINGLE_KEY };
```

ttnghia commented on cpp/benchmarks/join/join_heuristics.cu:
```diff
@@ -0,0 +1,191 @@
+    });
+  } else if constexpr (Method == heuristic_method::GROUPBY_MAX) {
+    // Approach 2: groupby count + max reduction
+    state.exec(nvbench::exec_tag::sync, [&](nvbench::launch& launch) {
```
```suggestion
    state.exec(nvbench::exec_tag::sync, [&](nvbench::launch&) {
```

ttnghia commented on cpp/include/cudf/join/key_remapping.hpp:
```diff
@@ -0,0 +1,158 @@
+   */
+  key_remapping(cudf::table_view const& build,
+                null_equality compare_nulls  = null_equality::EQUAL,
+                bool compute_metrics         = true,
```
cudf prefers to have boolean parameters replaced by `enum class enum_name { YES, NO };`.

```suggestion
enum class compute_metrics: bool { YES, NO };
....
...
key_remapping(....,
                compute_metrics metrics         = compute_metrics::YES,
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,682 @@
+    rmm::device_uvector<cudf::size_type> counts(build_num_rows, stream);
+    thrust::fill(rmm::exec_policy_nosync(stream), counts.begin(), counts.end(), 0);
+
+    rmm::device_scalar<cudf::size_type> d_distinct_count(0, stream);
```
Use `cudf::detail::device_scalar` which is backed by pinned host memory:
```suggestion
    cudf::detail::device_scalar<cudf::size_type> d_distinct_count{0, stream};
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,682 @@
+                                          counts.begin(),
+                                          counts.end(),
+                                          cudf::size_type{0},
+                                          thrust::maximum<cudf::size_type>{});
```
`thrust::` functional is deprecated and replaced by `cuda::`.

```suggestion
                                          cuda::maximum<cudf::size_type>{});
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,682 @@
+}  // namespace
+
+std::unique_ptr<cudf::column> key_remapping::remap_build_keys(
+  cudf::table_view const& keys,
```
Yeah, good point. If this function is to call on the same build table, we can get rid of the `keys` parameter and use the internally stored `_build` table.

ttnghia commented on cpp/tests/join/key_remapping_tests.cpp:
```diff
@@ -0,0 +1,776 @@
+  {
+    // Sort the keys+ids together by keys, then verify adjacent equal keys have equal ids
+    auto keys_with_ids_cols = keys.column(0).size() > 0 ? std::vector<cudf::column_view>{}
+                                                        : std::vector<cudf::column_view>{};
```
Something was missing here? Conditional values are the same.

ttnghia commented on java/src/main/java/ai/rapids/cudf/KeyRemapping.java:
```diff
@@ -0,0 +1,352 @@
+  public KeyRemapping(Table buildKeys, NullEquality nullEquality, boolean computeMetrics) {
+    this.nullEquality = nullEquality;
+    this.computeMetrics = computeMetrics;
+    Table buildTable = new Table(buildKeys.getColumns());
```
Should we implement `Table.incRefCount` to replace this step (creating a new table) instead?

ttnghia commented on java/src/main/java/ai/rapids/cudf/KeyRemapping.java:
```diff
@@ -0,0 +1,352 @@
+  }
+
+  /**
+   * Check if metrics (distinct_count, max_duplicate_count) were computed.
```
```suggestion
   * Check if metrics (distinctCount, maxDuplicateCount) were computed.
```

ttnghia commented on java/src/main/java/ai/rapids/cudf/KeyRemapping.java:
```diff
@@ -0,0 +1,352 @@
+    if (isClosed) {
+      throw new IllegalStateException("KeyRemapping is already closed");
+    }
+    return hasMetrics(cleaner.nativeHandle);
```
This calls down to the native function `has_metric` which basically just returns the same value of `computeMetrics`. Should we remove this JNI function?

ttnghia commented on java/src/main/java/ai/rapids/cudf/KeyRemapping.java:
```diff
@@ -0,0 +1,352 @@
+   * @throws IllegalArgumentException if keys has different number of columns than build table
+   * @throws CudfException if keys has different column types than build table
+   */
+  public ColumnVector remapBuildKeys(Table keys) {
```
Similar to my previous comment in https://github.com/rapidsai/cudf/pull/20826#discussion_r2658320532, if we want to remap the same build table, get rid of the `keys` parameter.

ttnghia commented on java/src/main/native/src/KeyRemappingJni.cpp:
```diff
@@ -0,0 +1,117 @@
+
+JNIEXPORT void JNICALL Java_ai_rapids_cudf_KeyRemapping_destroy(JNIEnv* env, jclass, jlong j_handle)
+{
+  JNI_TRY
```
```suggestion
{
  JNI_NULL_CHECK(env, j_handle, "handle is null", );
  JNI_TRY
```

revans2 commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,682 @@
+    auto preprocessed_probe =
+      cudf::detail::row::equality::preprocessed_table::create(probe_keys, stream);
+
+    if (cudf::detail::is_primitive_row_op_compatible(_build)) {
```
_build and probe have to be exactly the same or we throw an exception early on. There is no auto-up casting of int to long or anything like that.

revans2 commented:
@ttnghia and @abellina I think I have addressed all of your concerns, or I plan to look at the issues in a follow on PR. Please take another look

ttnghia commented on cpp/benchmarks/join/join_on_int32.cu:
```diff
@@ -0,0 +1,183 @@
+  return cudf::binary_operation(seq->view(),
+                                *divisor_scalar,
+                                cudf::binary_operator::DIV,
+                                cudf::data_type{cudf::type_id::INT32});
```
It seems that the data is monotonically increasing. This may lead to some bias in the benchmark, as equal keys stay very close to each other. I would recommend to shuffle them:
```
#include <thrust/shuffle.h>
#include <thrust/random.h>

auto seq  = cudf::sequence(num_rows, *init, *step, stream);

// Shuffle the generated sequence to reduce bias of data locality.
thrust::shuffle(rmm::exec_policy_nosync(stream), 
  seq->mutable_view().begin<int>(), 
  seq->mutable_view().end<int>(), 
  thrust::default_random_engine{});
```
(Ref: https://nvidia.github.io/cccl/thrust/api/group__shuffling_1ga5691add2f5ae54d226463c621eeb5455.html)

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,687 @@
+#include "join_common_utils.cuh"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
```
Unused header.
```suggestion
```

ttnghia commented on cpp/src/join/key_remapping.cu:
```diff
@@ -0,0 +1,687 @@
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <type_traits>
```
Unused header.
```suggestion
```

revans2 commented:
@ttnghia and @abellina I think I have addressed all of the review comments. I am going to go back and rerun the benchmarks just to be sure everything still looks good (now that the sort merge join change is merged in). I'll post the results when I am done.

ttnghia reviewed (approved):
🔥

revans2 commented:
/merge
